### PR TITLE
FSStore indexer error fix

### PR DESCRIPTION
**Pull request by andrewfulton9**
[Description of PR]

Adds tests for `Array` set up with `FSStore` and fixes the bug in  #644 

TODO:
* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [x] New/modified features documented in docs/tutorial.rst
* [x] Changes documented in docs/release.rst
* [x] AppVeyor and Travis CI passes
* [x] Test coverage is 100% (Coveralls passes)

### CODE REVIEW

**Comment by andrewfulton9** [comment]
Looks like `_set_selection` has the same bug as `_get_selection`. I'll go ahead and fix that as well and add that to the PR.

**Comment by andrewfulton9** [comment]
I am also having trouble understanding why `test_pickle` is failing. The raised error is that `FSMap` object has not attribute 'missing_exceptions', but I can't figure out why it wouldn't. I tried upgrading fsspec to 0.8.4 and that allows that test fo pass, but causes `test_storage.py::TestFSStore::test_s3` and `test_storage.py::TestFSStore::test_s3_complex` to fail. I assume to failure is due to the fixture, but I am not positive about that. Would it be prefered to upgrade fsspec and try to fix the s3 tests or try to figure out a way to fix `FSMap` issue? in the currently pinned version?

**Comment by Carreau** [comment]
From discussion with Andrew, the failure is/was dues to FSStore 0.8.3 not having self.missing_exceptions in the reduced; which was fixed in 0.8.4.

File: zarr/storage.py [hunk 1]
```diff
@@ -1113,7 +1113,9 @@ def rmdir(self, path=None):
             raise ReadOnlyError()
         store_path = self.dir_path(path)
         if self.fs.isdir(store_path):
-            self.fs.rm(store_path, recursive=True)
+            # sometimes doesn't delete on first try with S3
+            while self.fs.exists(store_path):
+                self.fs.rm(store_path, recursive=True)
```
Comment (Carreau) [inline zarr/storage.py:1118]: We may want to have exponential backoff with a timeout. But that can be done later in a subsequent PR.

**Review by Carreau — APPROVED** [review]

File: zarr/storage.py (hunk 1 above)
Comment (martindurant) [reply zarr/storage.py:1118]: This shouldn't be required - did you figure out what it going on? Perhaps just time for the metadata to sync?

Comment (andrewfulton9) [reply zarr/storage.py:1118]: @martindurant, I tried a `time.sleep` for various times (5, 10, 30 seconds) and that didn't seem to work. This is the only way I could figure out how to get it to pass without changing [test_s3_complex](https://github.com/zarr-developers/zarr-python/blob/06450d038aa4bd5502e1a72eee824c1d5f409494/zarr/tests/test_storage.py#L958-L999). Another way that I could get it to pass was by running :
```python
a = g.create_dataset("data", shape=(8, 8, 8), mode='w',
                             fill_value=-1, chunks=(1, 1, 1), overwrite=True)
a = g.create_dataset("data", shape=(8, 8, 8), mode='w',
                             fill_value=-1, chunks=(1, 1, 1), overwrite=True)
```
in place of the code [here](https://github.com/zarr-developers/zarr-python/blob/06450d038aa4bd5502e1a72eee824c1d5f409494/zarr/tests/test_storage.py#L997) and that also seemed to resolve the problem. For what it's worth, this test does't fail if fsspec is set to `0.8.3`, ony when its set to `0.8.4`, although at `0.8.3`, [test_pickle](https://github.com/zarr-developers/zarr-python/blob/06450d038aa4bd5502e1a72eee824c1d5f409494/zarr/tests/test_core.py#L759-L796) will fail.

**Pull request merged**